### PR TITLE
Add model_transform in create supervised trainer

### LINKED ISSUE #2837: Model Ouput Transform
## 🚀 Feature

The `*_training_step` implementations used by `create_supervised_trainer` expect only a single output from the model that's passed directly to the loss function, i.e.

        x, y = prepare_batch(batch, device=device, non_blocking=non_blocking)
        y_pred = model(x)
        loss = loss_fn(y_pred, y)

But I find that commonly models return more than predictions, i.e. they may return `logits` and `embeddings`, or `logits` and `loss` or perhaps a `dictionary`

It would be nice if you could pass a transform that would apply to the output from the model in order to extract the predictions in the form required by the loss function, i.e.

        x, y = prepare_batch(batch, device=device, non_blocking=non_blocking)
        outputs = model(x)
        y_pred = model_transform(outputs)
        loss = loss_fn(y_pred, y)

I have a current work-around - it required a lambda to handle the loss, and output_transform for each metric, i.e.

    loss = functools.partial(F.cross_entropy, label_smoothing=params['train']['label_smoothing'])
    trainer = create_supervised_trainer(model, 
                                        optimizer, 
                                        # model output is tuple, need to extract first element
                                        loss_fn=lambda output, target : loss(output[0], target), 
                                        prepare_batch=prepare_batch, 
                                        device=params['train']['device'])

    evaluator = create_supervised_evaluator(model,
                                            metrics={
                                                # model output is tuple, need to extract first element
                                                'accuracy': Accuracy(output_transform=lambda output : (output[0][0], output[1])),
                                                'loss': Loss(loss, output_transform=lambda output: (output[0][0], output[1]))},
                                            prepare_batch=prepare_batch,
                                            device=params['train']['device'])

It works fine but it was a bit fiddly to make sure I'd done it correctly, it would be a lot easier if I could just pass a single function to `create_supervised_trainer` i.e. `lambda output: output[0]`

### PR DESCRIPTION
**Pull request by guptaaryan16**
Fixes #2837 

Description: I have added the function for model_transform

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/engine/__init__.py [hunk 1]
```diff
@@ -48,6 +48,7 @@ def supervised_training_step(
     device: Optional[Union[str, torch.device]] = None,
     non_blocking: bool = False,
     prepare_batch: Callable = _prepare_batch,
+    model_transform: Callable[[torch.Tensor], torch.Tensor] = lambda output: output,
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:51]: I think the type hint should be a bit more generic: 
```suggestion
    model_transform: Callable[[Any], Any] = lambda output: output,
```

File: ignite/engine/__init__.py [hunk 2]
```diff
@@ -180,6 +187,7 @@ def update(engine: Engine, batch: Sequence[torch.Tensor]) -> Union[Any, Tuple[to
         x, y = prepare_batch(batch, device=device, non_blocking=non_blocking)
         with autocast(enabled=True):
             y_pred = model(x)
+            y_pred = model_transform(y_pred)
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:190]: Nit: let's rename variables a bit:
```diff
- y_pred = model(x)
- y_pred = model_transform(y_pred)
+ output = model(x)
+ y_pred = model_transform(output) 
```

File: ignite/engine/__init__.py [hunk 3]
```diff
@@ -64,6 +65,8 @@ def supervised_training_step(
             with respect to the host. For other cases, this argument has no effect.
         prepare_batch: function that receives `batch`, `device`, `non_blocking` and outputs
             tuple of tensors `(batch_x, batch_y)`.
+        model_transform: function that receives the y_pred from the model and convert it into the form as required
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:68]:
```suggestion
        model_transform: function that receives the output from the model and convert it into the form as required
```

**Review by vfdev-5 — COMMENTED** [review]
@guptaaryan16 thanks for the PR ! Looks good to me. Just few minor improvements and please add a test into https://github.com/pytorch/ignite/blob/master/tests/ignite/engine/test_create_supervised.py

**Comment by vfdev-5** [comment]
cc @david-waterworth if case if you would like to review this PR

File: tests/ignite/engine/test_create_supervised.py [hunk 4]
```diff
@@ -125,6 +125,113 @@ def _():
                 trainer.run(data)
 
 
+def _default_create_supervised_trainer_with_output_function(
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:128]: Instead of copy-pasting the code I think you can add a new arg to `_default_create_supervised_trainer` and `_test_create_supervised_trainer`. Then you can add a new line with the new arg activated into `test_create_supervised_trainer` to be sure to run the test

**Comment by vfdev-5** [comment]
@guptaaryan16 thanks for the update but let's improve test code

File: tests/ignite/engine/test_create_supervised.py (hunk 4 above)
Comment (guptaaryan16) [reply tests/ignite/engine/test_create_supervised.py:128]: ok sure I will do that. Should I make changes in the evaluator as well

File: ignite/engine/__init__.py [hunk 5]
```diff
@@ -560,6 +583,7 @@ def supervised_evaluation_step(
     device: Optional[Union[str, torch.device]] = None,
     non_blocking: bool = False,
     prepare_batch: Callable = _prepare_batch,
+    model_transform: Callable[[Any], Any] = lambda output: output,
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:586]: Actually, as you pointed out in another comment and I just remarked. I think `model_transform` is redundant here as `output_transform` already provide a similar functionality.

**Comment by vfdev-5** [comment]
@guptaaryan16 can you please update this PR to make it merged

**Comment by guptaaryan16** [comment]
Hey @vfdev-5 , sure I am making the changes by today or tomorrow EOD and I apologise for the delay in the same

File: tests/ignite/engine/test_create_supervised.py [hunk 6]
```diff
@@ -33,6 +33,7 @@ def _default_create_supervised_trainer(
     trace: bool = False,
     amp_mode: str = None,
     scaler: Union[bool, "torch.cuda.amp.GradScaler"] = False,
+    model_transform: Callable[[Any], Any] = lambda output: output,
```
[the remark this answers is not in the document]
Comment (guptaaryan16) [reply tests/ignite/engine/test_create_supervised.py:36]: ok I will do that

**Comment by guptaaryan16** [comment]
Hey @vfdev-5 can you give me an idea on how to update the API doc

File: tests/ignite/engine/test_create_supervised.py [hunk 7]
```diff
@@ -81,6 +84,77 @@ def _test_create_supervised_trainer(
...
+                assert not hasattr(state, "scaler")
+
+    else:
+        if Version(torch.__version__) >= Version("1.7.0"):
+            # This is broken in 1.6.0 but will be probably fixed with 1.7.0
+            with pytest.raises(RuntimeError, match=r"Expected all tensors to be on the same device"):
+                trainer.run(data)
+
+
+def _test_create_supervised_trainer_with_output_function(
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:132]: Hey @guptaaryan16 actually this still does not work as 1) this function is never called in the tests that are executing similar function `_test_create_supervised_trainer`, e.g. `test_create_supervised_trainer`.

The previous point https://github.com/pytorch/ignite/pull/2848/files#r1106001388 was about adding new tests case after this one:
https://github.com/pytorch/ignite/blob/4e66c46bb917e43787e3fa68a0a58a093af6dabb/tests/ignite/engine/test_create_supervised.py#L421

with `model_transform` non-default argument. Like this:
```python
_test_create_supervised_trainer(model_transform: lambda x: x[0])
```

**Comment by vfdev-5** [comment]
@guptaaryan16 I pushed this PR : https://github.com/guptaaryan16/ignite/pull/1 to your master branch on how to update the tests. Could you please merge this PR quickly to unblock this PR as we are planning the release. Thanks

**Comment by guptaaryan16** [comment]
Okay I am working in this issue and hope to complete this today

File: tests/ignite/engine/test_create_supervised.py [hunk 8]
```diff
@@ -1,14 +1,13 @@
 import os
 from importlib.util import find_spec
-from typing import Optional, Union
+from typing import Any, Callable, Optional, Union
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:3]:
```suggestion
from typing import Optional, Union
```

**Comment by vfdev-5** [comment]
> Okay I am working in this issue and hope to complete this today. Thanks for helping with the tests should I update the docs of the API

Thanks @guptaaryan16 ! Can you add `versionchanged` into the docstrings and also usage example can be helpful

**Comment by guptaaryan16** [comment]
@vfdev-5 I don't know why but to make the tests work I had to remove the bias variable in the tests as it gives a NoneType error

**Comment by vfdev-5** [comment]
> @vfdev-5 I don't know why but to make the tests work I had to remove the bias variable in the tests as it gives a NoneType error

I checked the tests locally and they were passing. Let's revert this change and see the exact error message

File: ignite/engine/__init__.py [hunk 9]
```diff
@@ -167,6 +169,8 @@ def supervised_training_step_amp(
     .. versionadded:: 0.4.5
     .. versionchanged:: 0.4.7
         Added Gradient Accumulation.
+    .. versionchanged:: 0.4.7
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:172]:
```suggestion
    .. versionchanged:: 0.4.11
```

File: ignite/engine/__init__.py [hunk 10]
```diff
@@ -335,6 +341,8 @@ def supervised_training_step_tpu(
     .. versionadded:: 0.4.5
     .. versionchanged:: 0.4.7
        Added Gradient Accumulation argument for all supervised training methods.
+    .. versionchanged:: 0.4.7
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:344]:
```suggestion
    .. versionchanged:: 0.4.11
```

File: ignite/engine/__init__.py [hunk 11]
```diff
@@ -515,6 +523,8 @@ def output_transform_fn(x, y, y_pred, loss):
 
     .. versionchanged:: 0.4.7
         Added Gradient Accumulation argument for all supervised training methods.
+    .. versionchanged:: 0.4.7
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:526]:
```suggestion
    .. versionchanged:: 0.4.11
```

**Review by vfdev-5 — APPROVED** [review]
Thanks @guptaaryan16
